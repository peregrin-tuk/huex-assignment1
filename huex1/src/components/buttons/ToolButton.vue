<template>
  <v-btn
    :fab="selected"
    :icon="!selected"
    :small="selected"
    color="white"
    @click="tool ? $store.commit('setActiveTool', tool) : null"
  >
    <v-icon
      :color="selected ? 'primary' : 'white'"
    >
      {{ icon }}
    </v-icon>
  </v-btn>
</template>

<script lang="ts">
import Vue from "vue";

export default Vue.extend({
  name: "ToolButton",
  props: {
    tool: String,
    icon: {
      type: String,
      required: true
    }
  },
  computed: {
    selected(): boolean {
      return this.$store.state.activeTool === this.tool
    }
  }
});
</script>

<style lang="scss" scoped>
</style>
