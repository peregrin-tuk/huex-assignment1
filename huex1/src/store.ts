/* eslint-disable no-unused-vars */
import Vue from 'vue'
import Vuex, {Store} from 'vuex'
import {firestoreAction, vuexfireMutations} from 'vuexfire'
import {db} from './database'
import firebase from "firebase";

Vue.use(Vuex)

const defaultCurrentBoard = {
  name: 'loading...',
  key: '',
  content: [],
  id: ''
}

const availableTools = ['select', 'brush', 'eraser', 'form', 'text']

const store: Store<any> = new Vuex.Store({
  state: {
    localStoredBoards: {},
    currentBoard: defaultCurrentBoard,
    activeTool: 'brush', // eg. 'brush' or 'form'
    toolProperties: {
      select: {},
      brush: {
        color: '#000000FF',
        size: 12
      },
      eraser: {
        size: 12
      },
      form: {
        strokeColor: '#000000FF',
        color: '#000000FF',
        formType: 'rectangle'
      },
      text: {
        color: '#000000FF',
        size: 12
      }
    },
    selectedBoardItem: null
  },

  getters: {
    getBoardName: state => state.currentBoard ? state.currentBoard.name : null,
    getBoardId: state => state.currentBoard ? state.currentBoard.id : null,
    getBoardContent: state => state.currentBoard ? state.currentBoard.content : null,
    activeTool: state => state.activeTool,
    activeToolHasColor: state => state.toolProperties[state.activeTool].color !== undefined,
    activeToolHasSize: state => state.toolProperties[state.activeTool].size !== undefined,
    getActiveToolSize: state => state.toolProperties[state.activeTool].size
  },

  mutations: {
    setActiveTool: (state, tool) => {
      availableTools.find((e) => e === tool) ?
        state.activeTool = tool : console.warn(`denied setting active tool to invalid value '${tool}', available values are${availableTools.map((tool) => ' ' + tool)} `)
    },
    resetCurrentBoardToDefaults: state => state.currentBoard = defaultCurrentBoard,
    setActiveToolColor(state, payload) {
      state.toolProperties[state.activeTool].color = payload
      // console.log(`commit: set ${state.activeTool} color to ${payload}`)
    },
    setActiveToolSize(state, payload) {
      state.toolProperties[state.activeTool].size = payload
      // console.log(`commit: set ${state.activeTool} size to ${payload}`)
    },
    addOrUpdateCurrentBoardToLocalBoards(state) {
      Vue.set(state.localStoredBoards, state.currentBoard.id, {
        name: state.currentBoard.name,
        timestamp: new Date()
      })
    },
    ...vuexfireMutations
  },

  actions: {
    bindCurrentBoard: firestoreAction(({bindFirestoreRef}, payload) => {
      return bindFirestoreRef('currentBoard',
        db.collection('boards').doc(payload), {reset: true})
        .then((e) => new Promise((resolve, reject) => {
          e ? resolve() : reject()
        }).catch((e) => {
          console.warn('Could not bind board because the board with the id ' + payload + ' could not be retrieved.')
        }))
    }),

    updateBoardName: firestoreAction((store, payload) => {
      return db.collection('boards')
        .doc(location.hash.substr(1))
        .update({name: payload})
        .then(e => {
          store.commit('addOrUpdateCurrentBoardToLocalBoards')
        })
        .catch(e => {
          console.error(e)
        })
    }),

    addNewBoard: firestoreAction(({bindFirestoreRef}) => {
      const ref = db.collection('boards').doc()
      return ref.set({
        name: 'New board',
        content: [],
        key: 'placeholder'
      }).then(e => {
        // db.collection('boards').doc(ref.id).get().then( (doc) => console.log(doc.data()))
        store.dispatch("bindCurrentBoard", ref.id)
      })
    }),

    addElementToCurrentBoardContent: firestoreAction((store, payload) => {
      return db.collection('boards')
        .doc(location.hash.substr(1))
        .update({
          content: firebase.firestore.FieldValue.arrayUnion(Object.assign({}, payload))
        })
        .then(e => {

        })
        .catch(e => {
          console.error(e)
        })
    }),
  }
});

export default store
