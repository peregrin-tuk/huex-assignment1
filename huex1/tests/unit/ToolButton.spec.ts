import {createLocalVue, shallowMount, mount} from '@vue/test-utils';
import ToolButton from "@/components/buttons/ToolButton.vue";
import Vuex from 'vuex'
import Vuetify from 'vuetify';
import store from "@/store.ts";

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.use(Vuetify)

describe('Toolbutton', () => {

  it('should match snapshot', () => {
    const wrapper = shallowMount(ToolButton, { 
      store,
      localVue,
      propsData: {
        tool: 'brush',
        icon: 'mdi-brush',
        selected: false
      }
    });

    expect(wrapper.html()).toMatchSnapshot();
  })

  it('contains a v-icon component', () => {
    const wrapper = mount(ToolButton, { 
      store, 
      localVue,
      propsData: {
        icon: 'mdi-icon'
      }
    });

    expect(wrapper.contains('.v-icon')).toBe(true);
  })

  it('renders an mdi icon based on the icon property', () => {
    const iconString = 'mdi-brush'
    const wrapper = mount(ToolButton, { 
      store,
      localVue,
      propsData: { icon: iconString }
    });

    const iconClasses = wrapper.find('.v-icon').classes();
    expect(iconClasses).toContain(iconString);
  })

  it('applies correct style classes when tool is selected', () => {
    const wrapper = mount(ToolButton, { 
      store,
      localVue,
      propsData: { 
        tool: 'brush',
        icon: 'mdi-icon'
      }
    });

    store.state.activeTool = 'brush'

    const buttonClasses = wrapper.find('.v-btn').classes();
    const iconClasses = wrapper.find('.v-icon').classes();

    expect(buttonClasses).toContain('v-btn--fab');
    expect(buttonClasses).toContain('v-size--small');
    expect(buttonClasses).not.toContain('v-btn--icon');

    expect(iconClasses).toContain('primary--text');
  })

  it('applies correct style classes when tool is not selected', () => {
    const wrapper = mount(ToolButton, { 
      store,
      localVue,
      propsData: {
        tool: 'eraser',
        icon: 'mdi-icon'
      }
    });

    store.state.activeTool = 'brush'

    let buttonClasses = wrapper.find('.v-btn').classes();
    let iconClasses = wrapper.find('.v-icon').classes();

    expect(buttonClasses).not.toContain('v-btn--fab');
    expect(buttonClasses).not.toContain('v-size--small');
    expect(buttonClasses).toContain('v-btn--icon');

    expect(iconClasses).toContain('white--text');
  })

})
