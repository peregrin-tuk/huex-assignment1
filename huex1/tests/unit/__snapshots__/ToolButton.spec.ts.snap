// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toolbutton should match snapshot 1`] = `<iconbutton-stub icon="mdi-brush"></iconbutton-stub>`;
